[14.0][l10n_br_account][IMP] composite account.move

rvalyi opened the pull request:
A single account.move can now eventually have different l10n_br_fiscal.document from its invoice_line_ids. This can be useful to import several CTe's or several NFSe's under the same account.move as these fiscal document allow only 1 item line and as this is usually done in other ERP's. At least we are reusing the _onchange_fiscal_document_line_id from this feature to properly complete the account.move.line with their proper onchanges.

This PR is used in the NFe importation PR https://github.com/OCA/l10n-brazil/pull/2781

rvalyi committed:
[IMP] composite account.move

rvalyi committed:
[REF] some cosmetic var renames

rvalyi committed:
[REF] All low level Documents menu

rvalyi committed:
[REF] all documents (account.move) menu

felipemotter commented:
Como ta a situação dessa proposta @rvalyi ? Já da pra testar de forma concreta?

rvalyi commented:
> Como ta a situação dessa proposta @rvalyi ? Já da pra testar de forma concreta?
> 

vou subir uns testes em breve...

rvalyi commented:
@felipemotter @antoniospneto @marcelsavegnago @renatonlima @mbcosta @mileo 

The PR is ready now for review now. I made e few changes and added a complete test suite (it's inside the test_account_move_lc.py file to avoid loading a fake chart of account another time, which would make the tests much slower).

As you can see, I added the fiscal_document_ids computed field with proper comments. Some extra checks are added in the case the user tries to perform an action on a move with several different fiscal documents.

Especially the fiscal view button now open the list of the related fiscal documents in the case there are more than 1.
For this, I changed the menu so there is now an "all low level Fiscal Documents" menu and action that will open the l10n_br_fiscal.document even when the l10n_br_account module is installed. This will also be useful to better track the webservice communication as pointed by @renatonlima and @antoniospneto.

Also now, in the UI, you can either:

1) create a new invoice line and select an existing fiscal document line to import it (the line will be populated with the onchange event):
![2024-03-21_07-35](https://github.com/OCA/l10n-brazil/assets/16926/03db3834-1a4d-46ee-b38a-c8ac6a168658)

2) or you can also import all the lines and other fiscal information from an existing l10n_br_fiscal.document right from the account.move form:
![2024-03-21_07-33](https://github.com/OCA/l10n-brazil/assets/16926/6ae130b7-aef5-4a0c-8f84-5e3eb10afc9e)

@felipemotter @antoniospneto you could probably use that to merge several CTe's into the same bill for instance. cc @douglascstd . Mileo you were also concerned with the need for such a feature for the NFSe's a few years ago...


Just to be clear: I could have made the NFe importation (https://github.com/OCA/l10n-brazil/pull/2781) work without this, but it would give me almost 2x more work as these two PRs share many concerns.

rvalyi commented:
Eu atualizei o PR de importação da NFe com esses 4 commits https://github.com/OCA/l10n-brazil/pull/2781#issuecomment-2012320590

antoniospneto commented on l10n_br_account/models/account_move_line.py:
```diff
@@ -151,10 +151,26 @@ def _inject_shadowed_fields(self, vals_list):
     @api.model_create_multi
     def create(self, vals_list):
         for values in vals_list:
+            if values.get("fiscal_document_line_id"):
+                fiscal_line_data = (
+                    self.env["l10n_br_fiscal.document.line"]
+                    .browse(values["fiscal_document_line_id"])
+                    .read(self._shadowed_fields())[0]
+                )
+                for k, v in fiscal_line_data.items():
+                    if isinstance(v, tuple):  # m2o
+                        values[k] = v[0]
+                    else:
+                        values[k] = v
+                continue
```
@rvalyi eu não certerza se compreendi bem aqui, aqui você faz o inverso do que é feito `_inject_shadowed_fields()` ?
ao invés de sincronizar o fiscal_line com o move_line, vc tá sincronizando o move_line com o fiscal_line isso ?

rvalyi replied:
eh isso sim. Nos testes desse PR eu testo isso de forma um pouco superficial, mas se vc testa de importar uma NFe em https://github.com/OCA/l10n-brazil/pull/2781 vai testar isso de forma bem funda. Inclusive eu penso botar outros testes depois em https://github.com/OCA/l10n-brazil/pull/2781 antes do merge...

Ai pelo sistema do _inherits a sincronização acontece de forma natural pelo framework (ou seja o account.move.line vai puxar os valoras fiscais da linha de documento fiscal pelo _inherits).
Os únicos campos que a gente tem que lidar são aqueles que são "shadowed" ou seja que não vão ser herdado mas que precisam ser duplicados no account.move.line. Esse código lida com essa copia desses 4 ou 5 campos.

felipemotter commented:
Irei revisar logo.

rvalyi commented:
@mileo não era isso que vc precisava aqui na sua RFC https://github.com/OCA/l10n-brazil/issues/894 ?
ai que tal testar...

mileo commented:
> @mileo não era isso que vc precisava aqui na sua RFC #894 ? ai que tal testar...

Sim, sobrando um tempo essa semana eu vou testar. Isso tem que fazer com bastante calma, mas é uma ótima contribuição, vai ajudar na questão do CT-e e outros documentos.

mileo commented:
@rvalyi pode fazer um rebase?

rvalyi commented:
> @rvalyi pode fazer um rebase?

fiz ontem, não tem mais conflito não (era apenas na versão)
CUIDADO se for fazer merge tem que ser com **nobump**.

rvalyi commented:
> @rvalyi pode fazer um rebase?

feito.
CUIDADO se for fazer merge depois tem que ser com **nobump**.

mileo reviewed (approved):
LGTM, do ponto de vista financeiro depois será necessário pensar em algumas melhorias, caso somente umas das linhas tenha sido cancelada.

- Seja removendo a linha ou dividindo o documento em dois, só com a parte não cancelada.
- Ou inserindo uma linha negativa para abater o valor.
- Ou até um reembolso da parte cancelada.

mileo commented:
> LGTM, do ponto de vista financeiro depois será necessário pensar em algumas melhorias, caso somente umas das linhas tenha sido cancelada.
> 
> * Seja removendo a linha ou dividindo o documento em dois, só com a parte não cancelada.
> * Ou inserindo uma linha negativa para abater o valor.
> * Ou até um reembolso da parte cancelada.

rvalyi commented:
a ideia é meio que definir uma "topologia" que funciona e que não conflita com a importação dos documentos fiscais (a NFe no caso). Depois da sim para melhorar muitas coisas em volta disso. Inclusive, vale a pena vc pensar bem se quer continuar fazer a NFSe como vc fez hoje somando todas linhas na mesma NFSe, pois poderia fazer uma NFSe por linha tb.

marcelsavegnago commented:
@rvalyi pode dar um rebase por favor ?

rvalyi commented:
> @rvalyi pode dar um rebase por favor ?

feito

marcelsavegnago reviewed (approved):
LGTM

rvalyi commented:
/ocabot merge nobump
